### DISCUSSION TITLE
As the command line code base grows larger, how do you encapsulate your commands, options and handlers?

### DISCUSSION THREAD

**Discussion by VigneshwarMoorthie** [dotnet/command-line-api#2181] [General]
My command line project is getting bigger and I am looking for an efficient way to separate out the commands and their handlers.

My plan was to create a class that inherits the command base class to create a command.

And then create another class which implements the command handler interface.

However I delcare the option in the command class and cannot retrieve it during the GetValueForOption method call in the command handler class?

**Comment by KalleOlaviNiemitalo** [comment]
I've been deriving classes from System.CommandLine.Command and handling all the System.CommandLine specifics in those. The constructor of each such class calls SetHandler with a delegate that retrieves the argument and option values from System.CommandLine and passes them to an Invoke method in a "handler" class that does not depend on System.CommandLine at all.

**Reply by VigneshwarMoorthie** [reply]
Doesn't the Invoke method require an InvocationContext object. How are you passing the arguments and options to the Invoke method? Does your handler class implement the ICommandHandler interface?

**Reply by KalleOlaviNiemitalo** [reply]
No it doesn't implement ICommandHandler.

The delegate does something like this

```
ConvertCommandHandler.Invoke(
    inputFile: context.GetValueForArgument(this.inputFileArgument),
    outputFile: context.GetValueForOption(this.outputFileOption));
```

I don't want to spread the System.CommandLine dependencies too much in the code base while the packages are in prerelease status.

**Comment by calacayir** [comment]
For this reason, we made DotMake.CommandLine which makes using System.CommandLine very easy and quick. Please check it out.
https://github.com/dotmake-build/command-line

```c#
/*
    Command hierarchy in below example is:
        
     TestApp
     └╴level-1
       └╴level-2
*/

[CliCommand(Description = "A root cli command with nested children")]
public class RootWithNestedChildrenCliCommand
{
    [CliOption(Description = "Description for Option1")]
    public string Option1 { get; set; } = "DefaultForOption1";

    [CliArgument(Description = "Description for Argument1")]
    public string Argument1 { get; set; } = "DefaultForArgument1";

    public void Run(CliContext context)
    {
        if (context.IsEmptyCommand())
            context.ShowHierarchy();
        else
            context.ShowValues();
    }

    [CliCommand(Description = "A nested level 1 sub-command")]
    public class Level1SubCliCommand
    {
        [CliOption(Description = "Description for Option1")]
        public string Option1 { get; set; } = "DefaultForOption1";

        [CliArgument(Description = "Description for Argument1")]
        public string Argument1 { get; set; }

        public void Run(CliContext context)
        {
            context.ShowValues();
        }

        [CliCommand(Description = "A nested level 2 sub-command")]
        public class Level2SubCliCommand
        {
            [CliOption(Description = "Description for Option1")]
            public string Option1 { get; set; } = "DefaultForOption1";

            [CliArgument(Description = "Description for Argument1")]
            public string Argument1 { get; set; }

            public void Run(CliContext context)
            {
                context.ShowValues();
            }
        }
    }
}
```
